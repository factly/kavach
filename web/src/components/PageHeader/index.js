import React, { useMemo } from 'react';
import { Link, useLocation } from 'react-router-dom';
import { PageHeader as AntPageHeader } from 'antd';
import routes from '../../config/routes';
import _ from 'lodash';

function PageHeader() {
  const location = useLocation();
  const pathSnippets = location.pathname.split('/').filter((i) => i);

  const breadcrumbItems = useMemo(() => {
    const urlBreadcrumbItems = pathSnippets.map((empty, index) => {
      const url = `/${pathSnippets.slice(0, index + 1).join('/')}`;
      const route = _.find(routes, { path: url });

      if (route) {
        return {
          path: route.path,
          breadcrumbName: route.title,
        };
      } else {
        return null;
      }
    });
    return [
      {
        path: '/',
        breadcrumbName: 'Home',
      },
    ].concat(_.filter(urlBreadcrumbItems));
  }, [pathSnippets]);

  const lastItem = useMemo(() => {
    return (
      _.find(routes, { path: breadcrumbItems[breadcrumbItems.length - 1].path }) || {
        path: '/',
        breadcrumbName: 'Home',
      }
    );
  }, [breadcrumbItems]);

  const itemRender = (route, params, routes, paths) => {
    const last = routes.indexOf(route) === routes.length - 1;
    if (last) {
      return <span>{route.breadcrumbName}</span>;
    }
    return <Link to={route.path}>{route.breadcrumbName}</Link>;
  };

  return (
    <AntPageHeader
      ghost={false}
      title={lastItem.title}
      breadcrumb={{ itemRender, routes: breadcrumbItems }}
    />
  );
}

export default PageHeader;
