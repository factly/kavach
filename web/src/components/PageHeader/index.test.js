import React from 'react';
import { useLocation } from 'react-router-dom';
import { BrowserRouter as Router } from 'react-router-dom';
import { PageHeader as AntPageHeader } from 'antd';
import _ from 'lodash';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';

import '../../matchMedia.mock';
import PageHeader from './index';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { act } from 'react-dom/test-utils';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useLocation: jest.fn().mockReturnValue({ pathname: '/users' }),
}));

describe('Page Header component', () => {
  let store;

  describe('snapshot testing', () => {
    it('should render component', () => {
      store = mockStore({});
      const tree = mount(
        <Provider store={store}>
          <Router>
            <PageHeader />
          </Router>
        </Provider>,
      );
      expect(tree).toMatchSnapshot();
    });
    it('should render component with route found', () => {
      useLocation.mockReturnValue({ pathname: '/users/create' });
      const state = {
        users: {
          loading: false,
        },
      };
      store = mockStore(state);
      const tree = mount(
        <Provider store={store}>
          <Router>
            <PageHeader />
          </Router>
        </Provider>,
      );
      expect(tree).toMatchSnapshot();
    });
  });
  describe('component testing', () => {
    let wrapper;

    it('should display breadcrumb for edit page', () => {
      useLocation.mockReturnValue({ pathname: '/applications/1/edit' });
      const state = {
        applications: {
          details: {
            1: { id: 1, name: 'Application' },
          },
          loading: false,
        },
      };
      store = mockStore(state);
      act(() => {
        wrapper = mount(
          <Provider store={store}>
            <Router>
              <PageHeader />
            </Router>
          </Provider>,
        );
      });
      expect(wrapper.find(AntPageHeader).props().title).toBe('Edit Application');
    });
    it('should not display breadcrumb for edit page when record not found', () => {
      useLocation.mockReturnValue({ pathname: '/applications/2/edit' });
      const state = {
        applications: {
          details: {
            1: { id: 1, name: 'Application' },
          },
          loading: false,
        },
      };
      store = mockStore(state);
      act(() => {
        wrapper = mount(
          <Provider store={store}>
            <Router>
              <PageHeader />
            </Router>
          </Provider>,
        );
      });
      const expectedRoutes = [
        { path: '/', breadcrumbName: 'Home' },
        { path: '/applications', breadcrumbName: 'Applications' },
        { path: '/applications/:id/edit', breadcrumbName: 'Application' },
      ];
      expect(wrapper.find(AntPageHeader).props().breadcrumb.routes).not.toEqual(expectedRoutes);
    });
    it('should not display breadcrumb for edit page when route not found', () => {
      useLocation.mockReturnValue({ pathname: '/application/2/edit' });
      const state = {
        application: {
          details: {
            1: { id: 1, name: 'Application' },
          },
          loading: false,
        },
      };
      store = mockStore(state);
      act(() => {
        wrapper = mount(
          <Provider store={store}>
            <Router>
              <PageHeader />
            </Router>
          </Provider>,
        );
      });
      const expectedRoutes = [{ path: '/', breadcrumbName: 'Home' }];
      expect(wrapper.find(AntPageHeader).props().breadcrumb.routes).toEqual(expectedRoutes);
    });
  });
});
