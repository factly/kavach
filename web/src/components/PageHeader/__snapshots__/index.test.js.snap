// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page Header component snapshot testing should render component 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <PageHeader />
    </Router>
  </BrowserRouter>
</Provider>
`;

exports[`Page Header component snapshot testing should render component with route found 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <PageHeader>
        <PageHeader
          breadcrumb={
            Object {
              "itemRender": [Function],
              "routes": Array [
                Object {
                  "breadcrumbName": "Home",
                  "path": "/",
                },
                Object {
                  "breadcrumbName": "Users",
                  "path": "/users",
                },
              ],
            }
          }
          ghost={false}
          title="Users"
        >
          <ResizeObserver
            onResize={[Function]}
          >
            <div
              className="ant-page-header has-breadcrumb"
            >
              <Breadcrumb
                itemRender={[Function]}
                routes={
                  Array [
                    Object {
                      "breadcrumbName": "Home",
                      "path": "/",
                    },
                    Object {
                      "breadcrumbName": "Users",
                      "path": "/users",
                    },
                  ]
                }
              >
                <div
                  className="ant-breadcrumb"
                >
                  <BreadcrumbItem
                    key="Home"
                    separator="/"
                  >
                    <span>
                      <span
                        className="ant-breadcrumb-link"
                      >
                        <Link
                          to="/"
                        >
                          <LinkAnchor
                            href="/"
                            navigate={[Function]}
                          >
                            <a
                              href="/"
                              onClick={[Function]}
                            >
                              Home
                            </a>
                          </LinkAnchor>
                        </Link>
                      </span>
                      <span
                        className="ant-breadcrumb-separator"
                      >
                        /
                      </span>
                    </span>
                  </BreadcrumbItem>
                  <BreadcrumbItem
                    key="users"
                    separator="/"
                  >
                    <span>
                      <span
                        className="ant-breadcrumb-link"
                      >
                        <span>
                          Users
                        </span>
                      </span>
                      <span
                        className="ant-breadcrumb-separator"
                      >
                        /
                      </span>
                    </span>
                  </BreadcrumbItem>
                </div>
              </Breadcrumb>
              <div
                className="ant-page-header-heading"
              >
                <div
                  className="ant-page-header-heading-left"
                >
                  <span
                    className="ant-page-header-heading-title"
                    title="Users"
                  >
                    Users
                  </span>
                </div>
              </div>
            </div>
          </ResizeObserver>
        </PageHeader>
      </PageHeader>
    </Router>
  </BrowserRouter>
</Provider>
`;
